## GET THE DATA & MAKE A DATAFRAME

In [1]:
#Install Libraries
import pandas as pd
import numpy as np

print ("ok, done with libraries")


ok, done with libraries


In [2]:
#Get the data from wikipedia
table = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
print ("got the data from wikipedia")


got the data from wikipedia


In [3]:
df=table[0]
df.head

<bound method NDFrame.head of     Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A  Downtown Toronto   
8        M8A      Not assigned   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
12       M2B      Not assigned   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
19       M7B      Not assigned   
20       M8B      Not assigned   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C      

## CLEAN THE DATA UP, REMOVE DUPLICATES & NOT ASSIGNED

In [4]:
#Join up Postcodes w/same neighborhoods
df =df.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(', '.join).reset_index()



In [5]:
#Drop Not Assigned Boroughs
df=df[df['Borough'] != 'Not assigned']
df


Postcode           Borough  \
1        M1B       Scarborough   
2        M1C       Scarborough   
3        M1E       Scarborough   
4        M1G       Scarborough   
5        M1H       Scarborough   
6        M1J       Scarborough   
7        M1K       Scarborough   
8        M1L       Scarborough   
9        M1M       Scarborough   
10       M1N       Scarborough   
11       M1P       Scarborough   
12       M1R       Scarborough   
13       M1S       Scarborough   
14       M1T       Scarborough   
15       M1V       Scarborough   
16       M1W       Scarborough   
17       M1X       Scarborough   
25       M2H        North York   
26       M2J        North York   
27       M2K        North York   
28       M2L        North York   
29       M2M        North York   
30       M2N        North York   
31       M2P        North York   
32       M2R        North York   
40       M3A        North York   
41       M3B        North York   
42       M3C        North York   
45       M3H        North York   
46       M3J        North York   
..       ...               ...   
102      M6C              York   
103      M6E              York   
104      M6G  Downtown Toronto   
105      M6H      West Toronto   
106      M6J      West Toronto   
107      M6K      West Toronto   
108      M6L        North York   
109      M6M              York   
110      M6N              York   
111      M6P      West Toronto   
112      M6R      West Toronto   
113      M6S      West Toronto   
120      M7A  Downtown Toronto   
132      M7R       Mississauga   
138      M7Y      East Toronto   
155      M8V         Etobicoke   
156      M8W         Etobicoke   
157      M8X         Etobicoke   
158      M8Y         Etobicoke   
159      M8Z         Etobicoke   
160      M9A      Queen's Park   
161      M9B         Etobicoke   
162      M9C         Etobicoke   
168      M9L        North York   
169      M9M        North York   
170      M9N              York   
171      M9P         Etobicoke   
172      M9R         Etobicoke   
175      M9V         Etobicoke   
176      M9W         Etobicoke   

                                         Neighbourhood  
1                                       Rouge, Malvern  
2               Highland Creek, Rouge Hill, Port Union  
3                    Guildwood, Morningside, West Hill  
4                                               Woburn  
5                                            Cedarbrae  
6                                  Scarborough Village  
7          East Birchmount Park, Ionview, Kennedy Park  
8                      Clairlea, Golden Mile, Oakridge  
9      Cliffcrest, Cliffside, Scarborough Village West  
10                         Birch Cliff, Cliffside West  
11   Dorset Park, Scarborough Town Centre, Wexford ...  
12                                   Maryvale, Wexford  
13                                           Agincourt  
14             Clarks Corners, Sullivan, Tam O'Shanter  
15   Agincourt North, L'Amoreaux East, Milliken, St...  
16                                     L'Amoreaux West  
17                                         Upper Rouge  
25                                   Hillcrest Village  
26                        Fairview, Henry Farm, Oriole  
27                                     Bayview Village  
28                            Silver Hills, York Mills  
29                             Newtonbrook, Willowdale  
30                                    Willowdale South  
31                                     York Mills West  
32                                     Willowdale West  
40                                           Parkwoods  
41                                     Don Mills North  
42                    Flemingdon Park, Don Mills South  
45     Bathurst Manor, Downsview North, Wilson Heights  
46                     Northwood Park, York University  
..                                                 ...  
102                                 Humewood-Cedarvale  
103           

In [6]:
df.shape

(103, 3)

In [7]:
#replace "not assigned" in neighborhood w/borough
df[df.Neighbourhood.str.contains("Not assigned")]

Postcode       Borough Neighbourhood
160      M9A  Queen's Park  Not assigned

In [8]:
df['Neighbourhood']=df['Neighbourhood'].replace('Not assigned', df['Borough'])

###DF has been cleaned up into 103 separate rows

In [10]:
df.shape

(103, 3)

# ADD LOCATIONS


## More Libraries

In [11]:
#install libraries
!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ca-certificates-2019.11.28 |       hecc5488_0         145 KB  conda-forge
    openssl-1.1.1d             |       h516909a_0         2.1 MB  conda-forge
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    certifi-2019.11.28         |           py36_0         149 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1.21.0-py_0       conda-forge

The following packages will be UPDATED:

    ca-

## GET GEO DATA & MERGE IT TO DATAFRAME

In [12]:
#read the data and inspect it
dfgeo=pd.read_csv("http://cocl.us/Geospatial_data")

In [13]:
dfgeo.head(5)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [14]:
dfgeo.shape

(103, 3)

In [15]:
#Reset the indexes to match up before merging
df.reset_index(drop=True, inplace=True)

In [16]:
df.head(5)

Postcode      Borough                           Neighbourhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

In [17]:
#merge the two into a new dataframe
dfnew = pd.merge(df, dfgeo, right_index=True, left_index=True)
dfnew.head(5)

Postcode      Borough                           Neighbourhood Postal Code  \
0      M1B  Scarborough                          Rouge, Malvern         M1B   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union         M1C   
2      M1E  Scarborough       Guildwood, Morningside, West Hill         M1E   
3      M1G  Scarborough                                  Woburn         M1G   
4      M1H  Scarborough                               Cedarbrae         M1H   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

In [18]:
#get rid of duplicate column
dfnew = dfnew.drop('Postal Code', 1)

In [85]:
dfnew.head(5)

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [21]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


### Let's Create a Map of Toronto

In [87]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, lng, borough, neighbourhood in zip(dfnew['Latitude'], dfnew['Longitude'], dfnew['Borough'], dfnew['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=3,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
map_toronto


### Add in Foursquare/Venue Data


In [131]:
# The code was removed by Watson Studio for sharing.

In [34]:
#Declare some more stuff:
LIMIT= 50
radius = 500
##Import some more stuff
import json
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans
print ("Done with this stuff")


Done with this stuff


In [35]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [49]:
#Run the function above

Toronto_venues = getNearbyVenues(names=dfnew['Neighbourhood'],
                                   latitudes=dfnew['Latitude'],
                                   longitudes=dfnew['Longitude']
                                  )


Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, 

In [50]:
Toronto_venues.shape

(1708, 7)

In [51]:
Toronto_venues.head()

Neighborhood  Neighborhood Latitude  \
0                          Rouge, Malvern              43.806686   
1  Highland Creek, Rouge Hill, Port Union              43.784535   
2  Highland Creek, Rouge Hill, Port Union              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.160497           Chris Effects Painting       43.784343   
2              -79.160497            Royal Canadian Legion       43.782533   
3              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
4              -79.188711                G & G Electronics       43.765309   

   Venue Longitude              Venue Category  
0       -79.199056        Fast Food Restaurant  
1       -79.163742  Construction & Landscaping  
2       -79.163085                         Bar  
3       -79.189914                 Pizza Place  
4       -79.191537           Electronics Store

### EXPLORE TORONTO VENUES DATA

In [45]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                               50   
Agincourt                                                               4   
Agincourt North, L'Amoreaux East, Milliken, Ste...                      3   
Albion Gardens, Beaumond Heights, Humbergate, J...                     11   
Alderwood, Long Branch                                                 10   
Bathurst Manor, Downsview North, Wilson Heights                        19   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      26   
Berczy Park                                                            50   
Birch Cliff, Cliffside West                                             4   
Bloordale Gardens, Eringate, Markland Wood, Old...                      8   
Brockton, Exhibition Place, Parkdale Village                           21   
Business Reply Mail Processing Centre 969 Eastern                      17   
CFB Toronto, Downsview East                                             4   
CN Tower, Bathurst Quay, Island airport, Harbou...                     18   
Cabbagetown, St. James Town                                            45   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  11   
Cedarbrae                                                               8   
Central Bay Street                                                     50   
Chinatown, Grange Park, Kensington Market                              50   
Christie                                                               18   
Church and Wellesley                                                   50   
Clairlea, Golden Mile, Oakridge                                        10   
Clarks Corners, Sullivan, Tam O'Shanter                                15   
Cliffcrest, Cliffside, Scarborough Village West                         3   
Commerce Court, Victoria Hotel                                         50   
Davisville                                                             37   
Davisville North                                                        8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     14   
...                                                                   ...   
Northwest                                                               2   
Northwood Park, York University                                         7   
Parkdale, Roncesvalles                                                 13   
Parkwoods                                                               3   
Queen's Park                                                           39   
Rosedale                                                                4   
Roselawn                                                                2   
Rouge, Malvern                                                          1   
Runnymede, Swansea                                                     39   
Ryerson, Garden District                                               50   
Scarborough Village                                                     1   
Silver Hills, York Mills                                                1   
St. James Town                                                         50   
Stn A PO Boxes 25 The Esplanade                                        50   
Studio District                                                        41   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             4   
The Beaches West, India Bazaar                                         19   
The Danforth West, Riverdale                                           41   
The Junction North, Runnymede                   

## Use One Hot encoding to breakdown by venue category

In [67]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe & move it over
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 


Toronto_onehot=Toronto_onehot[['Neighborhood'] + [col for col in Toronto_onehot.columns if col != 'Neighborhood']]

In [74]:

Toronto_onehot.head()

Neighborhood  Accessories Store  Airport  \
0                          Rouge, Malvern                  0        0   
1  Highland Creek, Rouge Hill, Port Union                  0        0   
2  Highland Creek, Rouge Hill, Port Union                  0        0   
3       Guildwood, Morningside, West Hill                  0        0   
4       Guildwood, Morningside, West Hill                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  ...  \
0                 0                    0             0  ...   
1                 0                    0             0  ...   
2                 0                    0             0  ...   
3                 0                    0             0  ...   
4                 0                    0             0  ...   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                   0                              0                 0   
1                   0                              0                 0   
2                   0                              0                 0   
3                   0                              0                 0   
4                   0                              0                 0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0            0                      0                0         0            0   
1            0                      0                0         0            0   
2            0                      0                0         0            0   
3            0                      0                0         0            0   
4            0                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 259 columns]

In [88]:
Toronto_onehot.shape

(1708, 259)

In [90]:
Toronto_group = Toronto_onehot.groupby('Neighborhood').mean().reset_index()

In [93]:
Toronto_group.head()

Neighborhood  Accessories Store  \
0                           Adelaide, King, Richmond                0.0   
1                                          Agincourt                0.0   
2  Agincourt North, L'Amoreaux East, Milliken, St...                0.0   
3  Albion Gardens, Beaumond Heights, Humbergate, ...                0.0   
4                             Alderwood, Long Branch                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0                 0.02           0.0  ...   
1               0.0                 0.00           0.0  ...   
2               0.0                 0.00           0.0  ...   
3               0.0                 0.00           0.0  ...   
4               0.0                 0.00           0.0  ...   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                           0.02               0.0   
1                 0.0                           0.00               0.0   
2                 0.0                           0.00               0.0   
3                 0.0                           0.00               0.0   
4                 0.0                           0.00               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0     0.000000                    0.0              0.0       0.0          0.0   
1     0.000000                    0.0              0.0       0.0          0.0   
2     0.000000                    0.0              0.0       0.0          0.0   
3     0.090909                    0.0              0.0       0.0          0.0   
4     0.000000                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 259 columns]

In [94]:
Toronto_group.shape


(98, 259)

In [95]:
num_top_venues = 3

for hood in Toronto_group['Neighborhood']:
    print("----"+hood+"----")
    temp = Toronto_group[Toronto_group['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
              venue  freq
0        Steakhouse  0.08
1              Café  0.06
2  Asian Restaurant  0.06


----Agincourt----
                       venue  freq
0  Latin American Restaurant  0.25
1                     Lounge  0.25
2             Breakfast Spot  0.25


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
         venue  freq
0   Playground  0.33
1  Coffee Shop  0.33
2         Park  0.33


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
           venue  freq
0    Pizza Place  0.18
1  Grocery Store  0.18
2       Pharmacy  0.09


----Alderwood, Long Branch----
          venue  freq
0   Pizza Place   0.2
1          Pool   0.1
2  Skating Rink   0.1


----Bathurst Manor, Downsview North, Wilson Heights----
           venue  freq
0    Coffee Shop  0.11
1       Pharmacy  0.05
2  Deli / Bodega  0.05


----Bayview Village----
                 venue  freq
0             

In [96]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [100]:
num_top_venues = 9

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_group['Neighborhood']

for ind in np.arange(Toronto_group.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_group.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond            Steakhouse   
1                                          Agincourt                Lounge   
2  Agincourt North, L'Amoreaux East, Milliken, St...            Playground   
3  Albion Gardens, Beaumond Heights, Humbergate, ...           Pizza Place   
4                             Alderwood, Long Branch           Pizza Place   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                       Café      Asian Restaurant    Seafood Restaurant   
1  Latin American Restaurant        Breakfast Spot        Clothing Store   
2                Coffee Shop                  Park           Yoga Studio   
3              Grocery Store              Pharmacy           Coffee Shop   
4                        Gym          Skating Rink          Dance Studio   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Hotel           Coffee Shop                   Bar   
1             Drugstore    Dim Sum Restaurant                 Diner   
2               Dog Run      Department Store          Dessert Shop   
3  Fast Food Restaurant            Beer Store        Sandwich Place   
4           Coffee Shop                  Pool                   Pub   

  8th Most Common Venue 9th Most Common Venue  
0             Gastropub           Salad Place  
1        Discount Store               Dog Run  
2    Dim Sum Restaurant                 Diner  
3   Fried Chicken Joint           Video Store  
4        Sandwich Place              Pharmacy

### CLUSTER NEIGHBORHOODS BY VENUE

In [101]:
# set number of clusters
kclusters = 8

Toronto_clustering = Toronto_group.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 0, 2, 0, 0, 4, 4, 4, 4, 4], dtype=int32)

In [110]:
Torontomerged = dfnew

# merge 
Torontomerged = Torontomerged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

Torontomerged.head() # check the last columns!

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Clusters  Cluster Labels 1st Most Common Venue  \
0 -79.194353       7.0             7.0  Fast Food Restaurant   
1 -79.160497       4.0             4.0                   Bar   
2 -79.188711       0.0             0.0     Electronics Store   
3 -79.216917       4.0             4.0           Coffee Shop   
4 -79.239476       4.0             4.0           Gas Station   

        2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Falafel Restaurant           Event Space  Ethiopian Restaurant   
1  Construction & Landscaping           Yoga Studio             Drugstore   
2          Mexican Restaurant           Pizza Place                   Spa   
3           Korean Restaurant           Yoga Studio      Doner Restaurant   
4                        Bank   Fried Chicken Joint  Caribbean Restaurant   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0   Empanada Restaurant     Electronics Store  Eastern European Restaurant   
1    Dim Sum Restaurant                 Diner               Discount Store   
2   Rental Car Location        Breakfast Spot                 Intersection   
3          Dessert Shop    Dim Sum Restaurant                        Diner   
4       Thai Restaurant    Athletics & Sports                       Bakery   

  8th Most Common Venue 9th Most Common Venue  
0   Dumpling Restaurant             Drugstore  
1               Dog Run      Doner Restaurant  
2        Medical Center        Discount Store  
3        Discount Store               Dog Run  
4      Hakka Restaurant   Dumpling Restaurant

# MAPPING TORONTO CLUSTERS

In [127]:
#dropNAN
Torontomerged2 = Torontomerged.dropna()

In [128]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)


# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(Torontomerged2['Latitude'], Torontomerged2['Longitude'], Torontomerged2['Neighbourhood'], Torontomerged2['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# DONE, THANK YOU FOR YOUR ATTENTION